### PR TITLE
[AIRFLOW-5073] Treat NULL as keep poking in SQL sensor

### PR DESCRIPTION
**Pull request by uncletoxa**
Change SQLSensor to treat NULL as fail. Add boolean allow_null
parameter to support legacy behavior passing NULL as success.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following:
  - https://issues.apache.org/jira/browse/AIRFLOW-5073

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
- switches NULL response to 'failed' result instead of 'success'
- adds `allow_true` parameter to preserve previous behaviour

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
- tests are extremely trivial, but I can add them if needed
### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

### CODE REVIEW

**Review by pgagnon — COMMENTED** [review]
I think you should add a test even though it's trivial, if only so people do not break your feature in the future.

**Comment by uncletoxa** [comment]
Bump

**Comment by dstandish** [comment]
Why not let the default for your param be the current behavior?

**Review by ashb — CHANGES_REQUESTED** [review]
Yup, please keep the default behaviour unchanged.

Oh you have. We don't need the entry in UPDATING then

**Comment by dstandish** [comment]
I think `allow_null=True` would preserve current behavior unless I'm missing something

**Comment by uncletoxa** [comment]
@ashb @dstandish don't you think changing the default behavior would be a good idea? I think 95% users would expect `NULL` result as non-valid response for sensor. We can change it now. All the existing users, who now rely on treating null as success, will receive news in `UPDATING.MD`.

File: airflow/sensors/sql_sensor.py
Comment (dstandish) [inline airflow/sensors/sql_sensor.py]: this could be made more clear.
Shouldn't it be "first cell is _not_ in ..."?

File: airflow/sensors/sql_sensor.py
Comment (dstandish) [inline airflow/sensors/sql_sensor.py]: it's not failure, right? it's just "not done", i.e. poke return false and keep retrying.

File: UPDATING.md
Comment (dstandish) [inline UPDATING.md]: >`'None'` is included in success criteria along with `0`, `'0'` and `''`. To preserve 

What you're actually doing is including `None` in "keep poking" criteria.

Previous behavior line is not worded clearly.  It excludes `None` from what?  Given what value?

Should be `allow_null` not `allow_true`

File: airflow/sensors/sql_sensor.py
Comment (dstandish) [inline airflow/sensors/sql_sensor.py]: > It will keep trying until success or failure criteria are met, or if the first cell is in (0, '0', '', 'None').

This could be made more clear.  Shouldn't it be "first cell is _not_ in ..."?

File: airflow/sensors/sql_sensor.py
Comment (dstandish) [inline airflow/sensors/sql_sensor.py:32]: > An allow_null parameter exclude 'None' results from failure criteria.

'None' was never part of failure criteria; `None` was implicitly part of the default _success_ criteria, since it was not on the "keep poking" list, and success is defined by negation.

**Comment by dstandish** [comment]
>  don't you think changing the default behavior would be a good idea? 

So, disclosure, I don't use sql sensor.  But here are my thoughts.
I agree the current design is quite awkward: `0` and `''` mean "not done", yet `NULL` means "done".

Having looked into this more, if anything I'd actually vote do _not_ add `allow_null` parameter; just add `None` to the "Not done" list, i.e. as though this were a bugfix.  The sensor is already confusing enough and if someone really needs `None` to mean "done" they can add a success callable e.g. `lambda x: str(x) == 'None'.  But I'm not that well versed in norms re changing behavior in a project like this.

In any case, I think it is misleading / confusing to say "treat null as fail".  In the context of sensors, I think fail means raise exception.  E.g. if the failure callback evaluates to True then an exception is raised.  What we're talking about here though is the "keep poking" list, i.e. whether a `None` results in a True or False, i.e. success or keep poking.  Maybe this is nit picky but I think something like "Do not interpret NULL as success" would be more accurate.

**Review by dstandish — CHANGES_REQUESTED** [review]
> don't you think changing the default behavior would be a good idea?

So, disclosure, I don't use sql sensor. But here are my thoughts.
I agree the current design is quite awkward: 0 and '' mean "not done", yet NULL means "done".

Having looked into this more, if anything I'd actually vote do _not_ add `allow_null` parameter; just add `None` to the "Not done" list, i.e. as though this were a bugfix. The sensor is already confusing enough and if someone really needs None to mean "done" they can add a success callable e.g. `lambda x: str(x) == 'None'`. But I'm not well versed in norms re changing behavior in a project like this.

In any case, I think it is misleading / confusing to say "treat null as fail". In the context of sensors, I think fail means raise exception. E.g. if the failure callback evaluates to True then an exception is raised. What we're talking about here though is the "keep poking" list, i.e. whether a `None` results in a True or False, i.e. success or keep poking. Maybe this is nit picky but I think something like "Do not interpret NULL as success" would be more accurate.

**Comment by ashb** [comment]
> don't you think changing the default behavior would be a good idea?

Yes, but this would be a quite likely breaking change in behaviour to existing sensors in deployed installs, so if you want this feature to be released before 2.0 then we need to keep the same behaviour.

**Comment by uncletoxa** [comment]
I strongly agree with this kind of fix:
> Having looked into this more, if anything I'd actually vote do _not_ add `allow_null` parameter; just add `None` to the "Not done" list, i.e. as though this were a bugfix. The sensor is already confusing enough and if someone really needs None to mean "done" they can add a success callable e.g. `lambda x: str(x) == 'None'`

Although I've changed default behavior to the legacy one to make it shipped to near releases.

**Comment by uncletoxa** [comment]
But still I'd like to make it less confusing in 2.0. @ashb could you guide me how to fix it for airflow 2.0?

**Comment by dstandish** [comment]
i could be wrong (i am also new to proj) but I think if you want non-breaking change in 1.10.x AND breaking change in 2.0, you will submit 2 PRs.  

The one that is a breaking change is based on master and goes to master; and non-breaking is based on v1-10-stable goes to v1-10-stable.

if you don't really care about getting it into 1.10.x then don't bother with the second PR.  update the ticket to set target version as 2.0.  i don't _think_ there is any convention in commits / PRs to  indicate "this is for 2.0 only" -- maybe check on dev slack channel

**Comment by uncletoxa** [comment]
to be honest, I'm quite lost. The issue I'm addressing was considered 
* as bug
* as a feature that should be deprecated
* as a feature that should be the default behavior 

I've tried to find some guides in [wiki](https://cwiki.apache.org) to understand deprecation policy and resolve confilct by myself, but found nothing relevant. Airflow 2.0 is also quite confusing. There is a roadmap, but no contribution guide, no timeline. Probably, I've chosen wrong pipeline to fix that kind of issue.

**Comment by ashb** [comment]
The easiest way for me as release manager is two PRs, both against master branch and targeting the same Jira ticket:

1. First PR: Add the flag, ensure current behaviour stays the same.
2. Second PR, after the first is merged, to change the default setting and add a note to UPDATING.md about the change in behaviour.

That way we can cherry-pick the first commit to the release branch but not the second.

**Comment by uncletoxa** [comment]
@ashb ok, then my PR is ready to be merged

**Review by ashb — APPROVED** [review]

File: airflow/sensors/sql_sensor.py
Comment (ashb) [inline airflow/sensors/sql_sensor.py:107]:
```suggestion
        return first_cell is not None and str(first_cell) not in ('0', '')
``

Comment (ashb) [reply airflow/sensors/sql_sensor.py:107]: (And although not  a new bug, this won't work if the type is floating point, that would give us 0.0 as a value)

**Pull request merged**
